arrow-libs/fx/arrow-fx-resilience/src/commonMain/kotlin/arrow/fx/resilience/Saga.kt

```diff
@@ -0,0 +1,178 @@
+ * data class Payment(val id: UUID, val orderId: UUID)
+ * ```
+ *
+ * The creation of an `Order` can only remain when a payment has been made. In SQL, you might run
+ * this inside a transaction, which can automatically roll back the creation of the `Order` when the
+ * creation of the Payment fails.
+ *
+ * When you need to do this across distributed services, or a multiple atomic references, etc. You
+ * need to manually facilitate the rolling back of the performed actions, or compensating actions.
+ *
+ * The [Saga] type, and [saga] DSL remove all the boilerplate of manually having to facilitate this
+ * with a convenient suspending DSL.
+ *
+ * ```kotlin
+ * data class Order(val id: UUID, val amount: Long)
+ * suspend fun createOrder(): Order = Order(UUID.randomUUID(), 100L)
+ * suspend fun deleteOrder(order: Order): Unit = println("Deleting $order")
+ *
+ * data class Payment(val id: UUID, val orderId: UUID)
+ * suspend fun createPayment(order: Order): Payment = Payment(UUID.randomUUID(), order.id)
+ * suspend fun deletePayment(payment: Payment): Unit = println("Deleting $payment")
+ *
+ * suspend fun Payment.awaitSuccess(): Unit = throw RuntimeException("Payment Failed")
+ *
+ * suspend fun main() {
+ *   saga {
+ *     val order = saga({ createOrder() }) { deleteOrder(it) }
+ *     val payment = saga { createPayment(order) }, ::deletePayment)
+ *     payment.awaitSuccess()
+ *   }.transact()
+ * }
+ * ```
+ */
+public typealias Saga<A> = suspend SagaScope.() -> A
+
+/** DSL that enables the [Saga] pattern in a `suspend` DSL. */
+@SagaDSLMarker
+public interface SagaScope {
+
+  /**
+   * Run an [action] to produce a value of type [A] and _install_ a [compensation] to undo the
+   * action.
+   */
+  @SagaDSLMarker
+  public suspend fun <A> saga(
+    action: suspend SagaActionStep.() -> A,
+    compensation: suspend (A) -> Unit,
+  ): A
+
+  /** Executes a [Saga] and returns its value [A] */
+  public suspend fun <A> Saga<A>.bind(): A = invoke(this@SagaScope)
+
+  /** Invoke a [Saga] and returns its value [A] */
+  public suspend operator fun <A> Saga<A>.invoke(): A = invoke(this@SagaScope)
+}
+
+/**
+ * The Saga builder which exposes the [SagaScope.bind]. The `saga` builder uses the suspension
+ * system to run actions, and automatically register their compensating actions.
+ *
+ * When the resulting [Saga] fails it will run all the required compensating actions, also when the
+ * [Saga] gets cancelled it will respect its compensating actions before returning.
+ *
+ * By doing so we can guarantee that any transactional like operations made by the [Saga] will
+ * guarantee that it results in the correct state.
+ */
+public inline fun <A> saga(noinline block: suspend SagaScope.() -> A): Saga<A> = block
+
+/** Create a lazy [Saga] that will only run when the [Saga] is invoked. */
+public fun <A> saga(
+  action: suspend SagaActionStep.() -> A,
+  compensation: suspend (A) -> Unit
+): Saga<A> = saga { saga(action, compensation) }
```

Comment:
This code is a bit puzzling. Is there any nice naming to separate the `saga` that takes `SagaActionStep` from the one that takes `SagaScope`?

Reply:
Hmm, I'm not sure I can think of a better name 😅
The goal here was similar then with `Resource`, but for `Resource` we not have a different name for the DSL version and the top-level _lazy_ version. I.e. for `Resource`:

```kotlin
val res: Resource<A> = resource(acquire, release)

val res2: Resource<A> = resource {
  install(acquire, release)
}
```

and for Saga we currently have:
```kotlin
val saga: Saga<A> = saga(action, compensate)
val saga2: Saga<A> = saga {
  saga(action, compensate)
}
```

The reason for `SagaActionStep`, and we do the same for `Resource` in `acquire`, is to disallow `bind()` being called from within the `action` or `acquire` step at compile-time.

So following the `Resource` pattern, then the DSL version probably needs a different name but I am drawing a black on what can be a good name for that operation 🤔